Hierarchical and structured config implementation: the initial pass

Zaid-Ajaj opened the pull request:
# Description

<!--- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. -->

This implements the initial pass of hierarchical and structured config which fixes #10602.

This changes the CLI such that configuration can now be defined at the _project_ level using a `config` block. The configuration values defined here are inherited by all the stacks and made available to the Pulumi program without having to duplicate values in every stack (hence hierarchical) and the values are also typed / structured. 

Example Project.yaml syntax:
```yaml
name: config-test
runtime: dotnet
config:
  instanceSize:
    type: string
    default: t3.micro
  instanceCount: 
    type: integer
    default: 5
```
This can also be rewritten using short-hand syntax and will be equivalent to the above
```yaml
name: config-test
runtime: dotnet
config:
  instanceSize: t3.micro
  instanceCount: 5
```
The complex types allowed for now are only arrays and nested arrays:
```yaml
name: config-test
runtime: dotnet
config:
  availabilityZones:
    type: array
    items: 
      type: string
    default: [us-east-1-atl-1a, us-east-1-chi-1a]
```


- Project-level configuration values that do not have a default value _MUST_ be defined at the stack level
- Stack configuration values are type-checked against their defined type in the project file i.e. Pulumi.yaml
- Short-hand syntax only accepts primitive values (no arrays for now)
- Accepted config types are a subset of a JSON schema where the property `type: string | integer | boolean | array` is expected. When `type: array` then a config block must also have property `items` which defines the type of array elements (can be nested)
- Running `pulumi config` will list the configuration values from the selected stack _AND_ the values inherited from the project
- After a successful `pulumi up` run using hierarchical config from the project, `pulumi config refresh` will write _ALL_ the used config back to the refreshed stack
- `pulumi config set/rm` only applies to the selected stack

## Checklist

<!--- Please provide details if the checkbox below is to be left unchecked. -->
- [x] I have added tests that prove my fix is effective or that my feature works
<!--- 
User-facing changes require a CHANGELOG entry.
-->
- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change
<!--
If the change(s) in this PR is a modification of an existing call to the Pulumi Service,
then the service should honor older versions of the CLI where this change would not exist.
You must then bump the API version in /pkg/backend/httpstate/client/api.go, as well as add
it to the service.
-->
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version
  <!-- @Pulumi employees: If yes, you must submit corresponding changes in the service repo. -->

Frassle commented on sdk/go/common/workspace/paths.go:
```diff
@@ -93,34 +93,19 @@ func DetectProjectPath() (string, error) {
 
 	fileName := fmt.Sprintf("%s.%s%s", ProjectFile, qnameFileName(stackName), filepath.Ext(projPath))
 
-	// Back compat: StackConfigDir used to be called Config.
```
We still need to support this. People could still have projects using "config: some_path". Ripping this out is another 4.0 change (maybe we should start leaving "TODO(4.0)" comments around for stuff like this)

Zaid-Ajaj replied:
Brought back support for defining `config: some_path` by rewriting it to `stackConfigDir` when it is a string. Erroring out when both are defined ✅

Frassle commented on sdk/go/common/workspace/project.json:
```diff
@@ -5,176 +5,290 @@
+      },
+      "config": {
+        "description": "A map of configuration keys to their types",
+        "type": "object",
```
```suggestion
        "type":  ["object", "string", "null"]
```
Still need to support the old string path here, and it's optional so a null is also valid.

Frassle commented on sdk/go/common/workspace/project.json:
```diff
@@ -5,176 +5,290 @@
+        ]
+      },
+      "config": {
+        "description": "A map of configuration keys to their types",
```
Add a bit to the description about how the "Config directory location relative to the location of Pulumi.yaml." is a deprecated use of this key.

Zaid-Ajaj replied:
Added that to the description

Frassle commented on sdk/go/common/workspace/project.json:
```diff
@@ -5,176 +5,290 @@
+              "$ref": "#/$defs/configItemsType"
+            }
+          },
+          "if": {
```
I wonder if a "oneOf" results in better errors here than "if" 🤔 
Not needed for this first pass but I might have a play with that idea later, so which is nicer UX.

Zaid-Ajaj commented on sdk/go/common/workspace/project.go:
```diff
@@ -176,29 +248,41 @@ func ValidateProject(raw interface{}) error {
 		return errors.New("project is missing a 'runtime' attribute")
 	}
 
+	project = RewriteShorthandConfigValues(project)
```
The project schema now validates short-hand config syntax ✅

Frassle replied:
Great but the rewrite here should still happen _after_ the Validate call below.
Validate the users input as is, then rewrite to the internal domain model.

Zaid-Ajaj commented on sdk/go/common/workspace/project.go:
```diff
@@ -122,7 +135,7 @@ type Project struct {
 
 	// Config has been renamed to StackConfigDir.
-	Config interface{} `json:"config,omitempty" yaml:"config,omitempty"`
+	Config map[string]ProjectConfigType `json:"config,omitempty" yaml:"config,omitempty"`
```
I've implemented project rewriting such that when we validate/marshal the project, if config was a string, it would already have been moved to `stackConfigDir` ✅

Zaid-Ajaj commented on pkg/cmd/pulumi/import.go:
```diff
@@ -492,7 +492,11 @@ func newImportCmd() *cobra.Command {
 
-			cfg, err := getStackConfiguration(ctx, s, sm)
+			cfg, err := getStackConfiguration(ctx, s, sm, StackConfigOptions{
+				// we don't need project config here
```
I wasn't sure in this case, so I changed this to inherit project-level config too ✅

Zaid-Ajaj commented:
> Can't loadProjectStack do this?

The reason `loadProjectStack` doesn't already apply inherited configuration values is because we do not always want to populate the configuration: especially for `pulumi config set` and `pulumi config rm` because when we save the config to disk, it will have the configuration from the project written out too and we don't want that.

We could remove the `project *Project` parameter but I think it is needed to remove the requirement of name-spacing configuration values with the project name: i.e. `project:config-key` is now required but I think it shouldn't be. Only `config-key` is needed and the `project` name should be auto detected

Frassle commented on pkg/cmd/pulumi/config.go:
```diff
@@ -39,6 +39,10 @@ import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 )
 
+type StackConfigOptions struct {
```
I think we can just get rid of this, I don't forsee any options here and applyProjectConfig should _always_ be true for now.

Frassle commented on pkg/cmd/pulumi/config.go:
```diff
@@ -741,12 +784,25 @@ type configValueJSON struct {
+	stackName := stack.Ref().Name().String()
+	// when listing configuration values
+	// also show values coming from the project
+	configError := workspace.ValidateStackConfigAndApplyProjectConfig(stackName, project, ps.Config)
```
I still think loadProjectStack should be handling this

Frassle commented on sdk/go/common/workspace/paths_test.go:
```diff
@@ -90,7 +90,7 @@ func TestProjectStackPath(t *testing.T) {
 		"name: some_project\ndescription: Some project\nruntime: nodejs\nconfig: stacksA\nstackConfigDir: stacksB\n",
 		func(t *testing.T, projectDir, path string, err error) {
 			assert.Error(t, err)
-			assert.Equal(t, "can not set `config` and `stackConfigDir`, remove the `config` entry", err.Error())
```
I think the original wording was better here, makes it clear that "stackConfigDir" is the one to keep.

Frassle replied:
Again I still prefer the original wording here telling the user to use "stackConfigDir"

Zaid-Ajaj replied:
The assert uses `Contains` to partially check the error message. The full error message is this:
> Should not use both config and stackConfigDir to define the stack directory. Use only stackConfigDir instead.

Frassle replied:
See no reason for the test to not assert that then.

Frassle commented on sdk/go/common/workspace/project_test.go:
```diff
@@ -62,6 +64,86 @@ func TestProjectValidation(t *testing.T) {
+		err.Error(),
+		"The default value specified for configuration key 'instanceSize' is not of the expected type 'integer'")
+
+	invalidValues := make([]interface{}, 0)
```
```suggestion
	invalidValues := [...]interface{} { "hello"} 
```
Inline array initialization is a little nicer than make/append.

Frassle commented on sdk/go/common/workspace/project_test.go:
```diff
@@ -125,44 +207,367 @@ func TestProjectLoadJSON(t *testing.T) {
 
 	// Test wrong type
-	_, err := writeAndLoad("\"hello\"")
+	_, err := loadProjectFromText(t, "\"hello\"")
```
That is a nicer function name!

Zaid-Ajaj commented:
> I still think loadProjectStack should be handling this

@Frassle  As discussed, because of `projectStack.Save()` we cannot always load the config into the project stack so we only apply where needed. 

> I think we can just get rid of `StackConfigOptions`, I don't forsee any options here and applyProjectConfig should always be true for now.

@Frassle Done ✅ 

> Would it be possible to collect these [missing keys error]? It would be a much better UX if we showed users all of the missing keys, instead of just the first one encountered.

@iwahbe Nice catch! Done ✅

I've addressed your comments in the latest update and added moaar tests, please have a second look when you have time 🙏 @Frassle @iwahbe

iwahbe commented on sdk/go/common/workspace/project.go:
```diff
@@ -97,6 +98,18 @@ type Plugins struct {
 	Analyzers []PluginOptions `json:"analyzers,omitempty" yaml:"analyzers,omitempty"`
 }
 
+type ProjectConfigItemsType struct {
+	Type  string                  `json:"type" yaml:"type"`
+	Items *ProjectConfigItemsType `json:"items" yaml:"items"`
+}
+
+type ProjectConfigType struct {
+	Type        string                  `json:"type" yaml:"type"`
+	Description string                  `json:"description" yaml:"description"`
+	Items       *ProjectConfigItemsType `json:"items" yaml:"items"`
+	Default     interface{}             `json:"default" yaml:"default"`
+}
+
```
We probably want some `omitempty`s here.

iwahbe commented on sdk/go/common/workspace/project.go:
```diff
@@ -143,8 +156,92 @@ type Project struct {
 
-func ValidateProject(raw interface{}) error {
-	// Cast any map[interface{}] from the yaml decoder to map[string]
+func isPrimitiveValue(value interface{}) (string, bool) {
```
We should include `float64` here.

Playing code gold, I think this would be better expressed as:
```go
switch value.(type) {
case bool:
    return "boolean", true
case int:
    return "integer", true
...
}
```

Frassle replied:
+1 to Ians suggestion to rewrite this as a switch

iwahbe commented on sdk/go/common/workspace/project.go:
```diff
@@ -226,6 +342,62 @@ func ValidateProject(raw interface{}) error {
 	return errs
 }
 
+func InferFullTypeName(typeName string, itemsType *ProjectConfigItemsType) string {
```
Pulumi YAML has a bunch of [type display logic](https://github.com/pulumi/pulumi-yaml/blob/e05b215177e550b3751c8c901eac1d0e4d6cd6e7/pkg/pulumiyaml/diags/types.go#L13). We should use it, probably necessitating moving it from pulumi/pulumi-yaml to sdk/go/common/???. I want to make sure we don't display the same type in different ways.

Zaid-Ajaj replied:
Looking at that display logic, it looks like Pulumi YAML prints array types as `List<T>` while here we have chosen to output `array<T>` instead. This is due to the fact that the type definition has `type: array` so it would look weird/unexpected that types would be printed as `List<T>`

Also the type display logic looks too generic for `schema.Type`, the config types here are a small subset of those so I would like to keep it simple as is

Frassle replied:
Yeh I'm going to lean to duplicating this display logic for now. Similar to comment below but config types != schema types, once we understand this domain better it might be something to clean up and unify.

iwahbe replied:
My concern is that users of Pulumi YAML will write
```yaml
config:
  foo:
    type: int
```
try to incorrectly use `foo`:

```yaml
resources:
  takesAListOfInt:
    properties:
      foo: ${foo}
```

and get an error:

```
foo: Cannot assign 'int' to 'List<int>'
```

If they give an a list of integers to the stack, they will get a different error:

```
Error: type mismatch between stack and project: found `array<int>` but expected `int`
```

Here the user need to understand that `List<int>` is equivalent to `array<int>` from (according to the user's perspective) the same program. I don't think that is a great experience. I choose `List` because it was what YAML used, but I'm not opposed to changing it. I just think it should be the same for the CLI and Pulumi YAML.

Frassle replied:
> Here the user need to understand that List<int> is equivalent to array<integer>

Same with any other language though right? C# would say `int[]`, Go would say `[]int`, python would say `List[int]`.

I fully expect us to have to iterate heavily on what we print here as we start exposing more complex object types for config, there may be types there that aren't actually usable by yaml so I don't think we want to necessarily tie to the same printing. What makes sense for config in general might be too verbose/complex for yaml if it's restricted.

iwahbe replied:
It is the same for every other language. I guess the difference is that we have the ability unify the Pulumi CLI and Pulumi YAML messaging. I don't think users will expect to need to context switch between the Pulumi CLI and Pulumi YAML. Pulumi YAML `configuration` accepts a heavily restricted subset of types but Pulumi YAML itself operates on the full `schema.Type`. I'm struggling to image a scenario where it would make sense to pass in a config value that it too complicated to be handed to a Pulumi resource (and thus Pulumi YAML).

This doesn't have to be a blocker. We can come back to error messages after the feature is more developed.

Frassle replied:
> We can come back to error messages after the feature is more developed.

Yeh totally happy to revisit this! I think once we think we've got type coverage for what we want to support at config level (and we know that's definitely more that what this first PR supports) we should look into both how the error messages look and unifying the use of those types and error messages in yaml.

iwahbe commented on sdk/go/common/workspace/project.go:
```diff
@@ -97,6 +98,18 @@ type Plugins struct {
 	Analyzers []PluginOptions `json:"analyzers,omitempty" yaml:"analyzers,omitempty"`
 }
 
+type ProjectConfigItemsType struct {
+	Type  string                  `json:"type" yaml:"type"`
+	Items *ProjectConfigItemsType `json:"items" yaml:"items"`
+}
```
Would it be possible to express this as a `schema.Type`? We already have a lot of logic to work on these, and we will need to translate them into schema types anyways for entry into the Pulumi YAML type checker.

Zaid-Ajaj replied:
It's possible to write a function to translate a config type into `schema.Type` but I wouldn't want to start with this right off the bat and only support what is needed because it would seem that a config type can be any `schema.Type` but that is not the case

Frassle replied:
Agree with Zaid here, config types != schema types, it will _probably_ end up being a subset but possibly a non-intersecting set.

iwahbe replied:
We will definitely need the `config type -> schema type` function when we incorporate this change into Pulumi YAML. We can wait for this change to settle before we unify.

Frassle replied:
That's a good point that this will probably have to be a subset otherwise there would be config values that YAML wouldn't be able to use, but I think given the subset nature of it I'm going to continue err'ing towards us not making it a schema.Type for now but it's own thing.

Frassle replied:
Actually see above, I think it might end up being non-intersecting still if we end up with config types that YAML doesn't support.

iwahbe commented on sdk/go/common/workspace/project.go:
```diff
@@ -226,6 +342,62 @@ func ValidateProject(raw interface{}) error {
+	return typeName
+}
+
+func ValidateConfigValue(typeName string, itemsType *ProjectConfigItemsType, value interface{}) bool {
```
I would appreciate a doc comment explaining what constitutes a valid config value.

Frassle commented on tests/integration/integration_test.go:
```diff
@@ -986,7 +986,6 @@ func TestDestroyStackRef(t *testing.T) {
 
 	e.RunCommand("pulumi", "up", "--skip-preview", "--yes")
 
-	e.CWD = os.TempDir()
```
Oh this was by design so that destroy didn't need a Pulumi.yaml file. We should see if we can maintain that support for now.

Zaid-Ajaj commented:
@Frassle I've fixed the issue about `pulumi destroy` requiring a project file, that is no longer the case and `TestDestroyStackRef` works nice as it did before. Please have another look when you have time 🙏

Frassle commented on sdk/go/common/workspace/project.json:
```diff
@@ -1,180 +1,307 @@
 {
-    "$schema": "https://json-schema.org/draft/2020-12/schema",
```
The indent change in this file makes diff compare very hard :( can we revert to 4 space tab?

Frassle commented on changelog/pending/20220922--cli-initial-mvp-config.yaml:
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: cli
+  description: Implement initial MVP for hierarchical and structured project configuration
```
```suggestion
  description: Implement initial MVP for hierarchical and structured project configuration.
```

Frassle commented on sdk/go/common/workspace/loaders.go:
```diff
@@ -121,33 +141,105 @@ type projectStackLoader struct {
+		return nil, err
+	}
+
+	if projectStackRaw == nil {
```
I am _very_ suspect of this. I assume this was to fix the "destory works without a Pulumi.yaml" but fixing it at this level means that any command using projects will try to use an empty project rather than failing validation, that feels wrong.

Frassle commented on sdk/go/common/workspace/loaders.go:
```diff
@@ -105,8 +106,27 @@ func (singleton *projectLoader) load(path string) (*Project, error) {
 		return nil, fmt.Errorf("could not validate '%s': %w", path, err)
 	}
 
+	// just before marshalling, we will rewrite the config values
```
I still think this rewrite should happen _after_ validation.
